### ISSUE TITLE
[feature] detect_api: Support detecting the version of a specific compiler

### ISSUE THREAD

**Issue by jwillikers** [conan-io/conan#15978]
### What is your suggestion?

I would like to use more of the `detect_api` functionality in my company's profiles as it's more flexible, less configuration, and less-error prone than managing all of the profiles manually. One limitation is that obtaining the version of the compiler is only supported for the default compiler and not non-default compilers. We have `gcc` and `clang` profiles and would like to use the `detect_api` to automatically pull the version for each of these compilers specifically. This might end up looking something like the following snippet.

```
[settings]
compiler = clang
compiler.version = {{ detect_api.default_compiler_version("clang", detect_api.detect_compiler_version("clang"))  }}
```

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by jcar87** [comment]
Hi @jwillikers  - thanks for this feature request.

Bear in mind that for compilers that are _not_ the system default, the profile needs to indicate _where_ this compiler is located, otherwise Conan assumes that other tools just run with the system defaults as well.  This is addressed by defining the `tools.build:compiler_executables` conf, or with `[buildenv]` vars where appropriate.

At the moment the `detect_default_compiler()` functionality is biased towards locating a default given platform priorities, or influenced by the `CC`/`CXX` environment variables - I suppose the functionality that is missing here is: if we are given a compiler executable, please fill in the rest for me. 

Or alternatively, the ability to alter the priority of `detect_api.detect_default_compiler()` - at the moment it can already be influenced by externally to the profile file (by defining the CC environment variable when invoking Conan - this is harder to guarantee for dev machines, but it's useful for Docker images for example). However here as well we would need to ensure the profile defines the conf or env vars when used.


Otherwise, for the first option, I think the profile could look something along the lines of:
```
[settings]
{% set c_compiler = '/path/to/clang' %}
{% set cxx_compiler = '/path/to/clang++' %}
{% set compiler, version  = detect_api.detect_compiler_settings(cxx_compiler) %}
compiler = {{compiler}} 
compiler.version = {{version}}
# (derive other settings)


[conf]
tools.build:compiler_executables={ 'c':{{c_compiler}},'cxx':{{cxx_compiler}} }
```

**Comment by memsharded** [comment]
Proposing new ``detect_api.detect_xxx_compiler`` apis in https://github.com/conan-io/conan/pull/16092

**Comment by ErniGH** [comment]
After the merge of #16092 an example of a profile would be the following:
```
[settings]
{% set c_compiler = 'clang' %}
{% set cxx_compiler = 'clang++' %}
{% set compiler, version, _ = detect_api.detect_clang_compiler(c_compiler) %}

compiler = {{compiler}}
compiler.version = {{detect_api.default_compiler_version(compiler, version)}}
compiler.libcxx = {{detect_api.detect_libcxx(compiler, version)}}
compiler.cppstd = {{detect_api.detect_cppstd(compiler, version)}}

os=Linux
arch=armv8

[conf]
tools.build:compiler_executables={ 'c':{{c_compiler}},'cxx':{{cxx_compiler}} }
```

**Comment by jwillikers** [comment]
> After the merge of #16092 an example of a profile would be the following:
> 
> ```
> [settings]
> {% set c_compiler = 'clang' %}
> {% set cxx_compiler = 'clang++' %}
> {% set compiler, version, _ = detect_api.detect_clang_compiler(c_compiler) %}
> 
> compiler = {{compiler}}
> compiler.version = {{detect_api.default_compiler_version(compiler, version)}}
> compiler.libcxx = {{detect_api.detect_libcxx(compiler, version)}}
> compiler.cppstd = {{detect_api.detect_cppstd(compiler, version)}}
> 
> os=Linux
> arch=armv8
> 
> [conf]
> tools.build:compiler_executables={ 'c':{{c_compiler}},'cxx':{{cxx_compiler}} }
> ```

Thanks! That's exactly what I'm looking for!

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/16092 merged for next Conan 2.3
